File: sdkwork-rtc-sdk-swift/Sources/RtcSdk/RtcProviderExtensionCatalog.swift
public struct RtcProviderExtensionCatalogEntry {
    public let extensionKey: String
    public let providerKey: String
    public let displayName: String
    public let surface: String
    public let access: String
    public let status: String
}

public enum RtcProviderExtensionCatalog {
    public static let recognizedSurfaces: [String] = [
        "control-plane",
        "runtime-bridge",
        "cross-surface",
    ]

    public static let recognizedAccessModes: [String] = [
        "unwrap-only",
        "extension-object",
    ]

    public static let recognizedStatuses: [String] = [
        "reference-baseline",
        "reserved",
    ]

    public static let entries: [RtcProviderExtensionCatalogEntry] = [
        .init(extensionKey: "volcengine.native-client", providerKey: "volcengine", displayName: "Volcengine Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reference-baseline"),
        .init(extensionKey: "aliyun.native-client", providerKey: "aliyun", displayName: "Aliyun Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reference-baseline"),
        .init(extensionKey: "tencent.native-client", providerKey: "tencent", displayName: "Tencent Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reference-baseline"),
        .init(extensionKey: "agora.native-client", providerKey: "agora", displayName: "Agora Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "zego.native-client", providerKey: "zego", displayName: "ZEGO Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "livekit.native-client", providerKey: "livekit", displayName: "LiveKit Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "twilio.native-client", providerKey: "twilio", displayName: "Twilio Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "jitsi.native-client", providerKey: "jitsi", displayName: "Jitsi Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "janus.native-client", providerKey: "janus", displayName: "Janus Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
        .init(extensionKey: "mediasoup.native-client", providerKey: "mediasoup", displayName: "mediasoup Native Client", surface: "runtime-bridge", access: "unwrap-only", status: "reserved"),
    ]

public static func getRtcProviderExtensionCatalog() -> [RtcProviderExtensionCatalogEntry] {
        entries
    }

    public static func getRtcProviderExtensionDescriptor(_ extensionKey: String) -> RtcProviderExtensionCatalogEntry? {
        entries.first { $0.extensionKey == extensionKey }
    }

    public static func getRtcProviderExtensionsForProvider(_ providerKey: String) -> [RtcProviderExtensionCatalogEntry] {
        entries.filter { $0.providerKey == providerKey }
    }

    public static func getRtcProviderExtensions(_ extensionKeys: [String]) -> [RtcProviderExtensionCatalogEntry] {
        var resolved: [RtcProviderExtensionCatalogEntry] = []
        for extensionKey in extensionKeys {
            if let entry = getRtcProviderExtensionDescriptor(extensionKey) {
                resolved.append(entry)
            }
        }

        return resolved
    }

    public static func hasRtcProviderExtension(_ extensionKeys: [String], _ extensionKey: String) -> Bool {
        extensionKeys.contains(extensionKey) && getRtcProviderExtensionDescriptor(extensionKey) != nil
    }

}

File: sdkwork-rtc-sdk-kotlin/src/main/kotlin/com/sdkwork/rtc/metadata/RtcProviderExtensionCatalog.kt
package com.sdkwork.rtc.metadata

data class RtcProviderExtensionCatalogEntry(
    val extensionKey: String,
    val providerKey: String,
    val displayName: String,
    val surface: String,
    val access: String,
    val status: String,
)

object RtcProviderExtensionCatalog {
    val recognizedSurfaces: List<String> = listOf(
        "control-plane",
        "runtime-bridge",
        "cross-surface",
    )

    val recognizedAccessModes: List<String> = listOf(
        "unwrap-only",
        "extension-object",
    )

    val recognizedStatuses: List<String> = listOf(
        "reference-baseline",
        "reserved",
    )

    val entries: List<RtcProviderExtensionCatalogEntry> = listOf(
        RtcProviderExtensionCatalogEntry("volcengine.native-client", "volcengine", "Volcengine Native Client", "runtime-bridge", "unwrap-only", "reference-baseline"),
        RtcProviderExtensionCatalogEntry("aliyun.native-client", "aliyun", "Aliyun Native Client", "runtime-bridge", "unwrap-only", "reference-baseline"),
        RtcProviderExtensionCatalogEntry("tencent.native-client", "tencent", "Tencent Native Client", "runtime-bridge", "unwrap-only", "reference-baseline"),
        RtcProviderExtensionCatalogEntry("agora.native-client", "agora", "Agora Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("zego.native-client", "zego", "ZEGO Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("livekit.native-client", "livekit", "LiveKit Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("twilio.native-client", "twilio", "Twilio Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("jitsi.native-client", "jitsi", "Jitsi Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("janus.native-client", "janus", "Janus Native Client", "runtime-bridge", "unwrap-only", "reserved"),
        RtcProviderExtensionCatalogEntry("mediasoup.native-client", "mediasoup", "mediasoup Native Client", "runtime-bridge", "unwrap-only", "reserved"),
    )

fun getRtcProviderExtensionCatalog(): List<RtcProviderExtensionCatalogEntry> = entries

    fun getRtcProviderExtensionDescriptor(extensionKey: String): RtcProviderExtensionCatalogEntry? =
        entries.firstOrNull { it.extensionKey == extensionKey }

    fun getRtcProviderExtensionsForProvider(providerKey: String): List<RtcProviderExtensionCatalogEntry> =
        entries.filter { it.providerKey == providerKey }

    fun getRtcProviderExtensions(extensionKeys: List<String>): List<RtcProviderExtensionCatalogEntry> =
        extensionKeys.mapNotNull(::getRtcProviderExtensionDescriptor)

    fun hasRtcProviderExtension(extensionKeys: List<String>, extensionKey: String): Boolean =
        extensionKeys.contains(extensionKey) && getRtcProviderExtensionDescriptor(extensionKey) != null

}
